### PR TITLE
CBG-415: Improve raw endpoint for support utility

### PR DESCRIPTION
**Pull request by JRascagneres**

### CODE REVIEW

File: db/change_cache.go [hunk 1]
```diff
@@ -392,7 +392,7 @@ func (c *changeCache) DocChanged(event sgbucket.FeedEvent) {
 
 		if syncData == nil || !isSGWrite {
 			if c.context.autoImport {
-				// If syncData is nil, or if this was not an SG write, attempt to import
+				// If SyncData is nil, or if this was not an SG write, attempt to import
```
Comment (bbrks) [inline db/change_cache.go]:
```suggestion
				// If syncData is nil, or if this was not an SG write, attempt to import
```

File: db/design_doc.go [hunk 2]
```diff
@@ -340,7 +340,7 @@ func checkExistingDDocs(bucket base.Bucket) bool {
 
 func installViews(bucket base.Bucket) error {
 
-	// syncData specifies the path to Sync Gateway sync metadata used in the map function -
+	// SyncData specifies the path to Sync Gateway sync metadata used in the map function -
```
Comment (bbrks) [inline db/design_doc.go]:
```suggestion
	// syncData specifies the path to Sync Gateway sync metadata used in the map function -
```

File: rest/admin_api.go [hunk 3]
```diff
@@ -317,10 +317,27 @@ func (h *handler) handleActiveTasks() error {
 func (h *handler) handleGetRawDoc() error {
 	h.assertAdminOnly()
 	docid := h.PathVar("docid")
+	includeDoc := h.getBoolQuery("include_doc")
+	redact := h.getOptBoolQuery("redact", true)
+	salt := h.getQuery("salt")
```
Comment (bbrks) [inline rest/admin_api.go:323]: Should the default salt be randomised on each call? I think that's how sgcollect_info does it for an undefined salt.

File: rest/admin_api.go [hunk 4]
```diff
@@ -317,10 +317,27 @@ func (h *handler) handleActiveTasks() error {
...
+
+	if redact {
+		h.response.Write([]byte(`{"_sync":` + doc.SyncData.HashRedact(salt)))
+	} else {
+		h.response.Write([]byte(`{"_sync":`))
+		h.addJSON(doc.SyncData)
 	}
+
+	if includeDoc {
+		h.response.Write([]byte(`, "doc":`))
```
Comment (bbrks) [inline rest/admin_api.go]: Nesting the doc body under `"doc"` feels a little wrong to do here, given the raw document is typically at the root. Perhaps we could iterate over each element in doc.Body() and append it to the JSON instead?

File: base/util.go [hunk 5]
```diff
@@ -1135,3 +1136,10 @@ func (ab *AtomicBool) Set(flag bool) {
 func (ab *AtomicBool) IsTrue() bool {
 	return atomic.LoadInt32(&ab.value) == 1
 }
+
+func Sha1HashString(string string, salt string) string {
+	h := sha1.New()
+	h.Write([]byte(string + salt))
```
Comment (bbrks) [inline base/util.go]: The sgcollect_info version has the salt as the first element, and then the data. We'll need to try and make sure these things hash the same way.

```suggestion
	h.Write([]byte(salt + str))
```

File: base/util.go [hunk 6]
```diff
@@ -1135,3 +1136,10 @@ func (ab *AtomicBool) Set(flag bool) {
 func (ab *AtomicBool) IsTrue() bool {
 	return atomic.LoadInt32(&ab.value) == 1
 }
+
+func Sha1HashString(string string, salt string) string {
```
Comment (bbrks) [inline base/util.go]:
```suggestion
func Sha1HashString(str string, salt string) string {
```

**Review by bbrks — CHANGES_REQUESTED** [review]

File: db/document.go [hunk 7]
```diff
@@ -76,12 +76,60 @@ type syncData struct {
...
+		RecentSequences: doc.RecentSequences,
+		History:         RevTree{},
+		Channels:        channels.ChannelMap{},
+		Access:          UserAccessMap{},
+		RoleAccess:      UserAccessMap{},
+		Expiry:          doc.Expiry,
+		Cas:             doc.Cas,
+		Crc32c:          doc.Crc32c,
+		TombstonedAt:    doc.TombstonedAt,
+		Attachments:     doc.Attachments,
```
Comment (bbrks) [inline db/document.go]: Attachment names are probably considered UserData actually after reviewing this list.

**Review by bbrks — CHANGES_REQUESTED** [review]

File: rest/doc_api_test.go [hunk 8]
```diff
@@ -106,7 +106,7 @@ func TestDocumentNumbers(t *testing.T) {
 			}
 
 			// Check channel assignment
-			getRawResponse := rt.SendAdminRequest("GET", fmt.Sprintf("/db/_raw/%s", test.name), "")
+			getRawResponse := rt.SendAdminRequest("GET", fmt.Sprintf("/db/_raw/%s?redact=false", test.name), "")
```
Comment (adamcfraser) [inline rest/doc_api_test.go:109]: Add some tests for the new functionality.

File: db/document.go [hunk 9]
```diff
@@ -76,12 +76,64 @@ type syncData struct {
...
+			for existingChanKey := range v.Channels {
+				revInfo.Channels.Add(base.Sha1HashString(existingChanKey, salt))
+			}
+		}
+
+		syncData.History.addRevision(k, revInfo)
+	}
+
+	for k, v := range doc.Access {
+		syncData.Access[base.Sha1HashString(k, salt)] = v
```
Comment (adamcfraser) [inline db/document.go]: I think 'v' is a channel name here, which would need to get hashed as well. (same in the RoleAccess case below)

File: db/document.go [hunk 10]
```diff
@@ -76,12 +76,64 @@ type syncData struct {
 	removedRevisionBodyKeys map[string]string // keys of non-winning revisions that have been removed (and so may require deletion), indexed by revID
 }
 
+func (doc *SyncData) HashRedact(salt string) SyncData {
+
+	syncData := SyncData{
```
Comment (adamcfraser) [inline db/document.go:83]: Maybe just stick a comment above this describing what's being copied and why.

File: db/document.go [hunk 11]
```diff
@@ -76,12 +76,64 @@ type syncData struct {
 	removedRevisionBodyKeys map[string]string // keys of non-winning revisions that have been removed (and so may require deletion), indexed by revID
 }
 
+func (doc *SyncData) HashRedact(salt string) SyncData {
```
Comment (adamcfraser) [inline db/document.go]: Using 'doc' for *SyncData as it's usually used for 'document' in the code - would make it clearer to just use 'sd' or something.

File: db/document.go [hunk 12]
```diff
@@ -804,10 +856,14 @@ func (doc *document) MarshalWithXattr() (data []byte, xdata []byte, err error) {
...
+	xdata, err = json.Marshal(doc.SyncData)
 	if err != nil {
-		return nil, nil, pkgerrors.WithStack(base.RedactErrorf("Failed to MarshalWithXattr() doc syncData with id: %s.  Error: %v", base.UD(doc.ID), err))
+		return nil, nil, pkgerrors.WithStack(base.RedactErrorf("Failed to MarshalWithXattr() doc SyncData with id: %s.  Error: %v", base.UD(doc.ID), err))
 	}
 
 	return data, xdata, nil
 }
+
+func (doc *document) GetSyncData() SyncData {
```
Comment (adamcfraser) [inline db/document.go]: Is this accessor needed, now that the struct is public?

**Review by adamcfraser — CHANGES_REQUESTED** [review]

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**